core/cap-0007.md

```diff
@@ -0,0 +1,150 @@
+## Preamble
+
+```
+CAP: 0007
+Title: Deterministic Account Creation
+Author: Jeremy Rubin
+Status: Draft
+Created: 2018-11-07
+Discussion: https://github.com/stellar/stellar-protocol/issues/99
+Protocol version: TBD
+```
+
+## Simple Summary
+We introduce CreateDeterministicAccount operation which creates an account that
+could only be created by the transaction creating it, allowing it to have a
+known starting sequence number and initialization.
+
+## Abstract
+
+
+Deterministic accounts are a critical infrastructure piece for building reliable contracts for three key reasons:
+
+1) Predictable sequence numbers
+2) Never-fail creation
+3) Known initialized state
+
+We can create deterministic accounts with minimal changes to Stellar by using Key Tweaking to derive the account key.
+
+
+## Motivation
+
+Currently, when a contract must create an account as a part of it's settlement (such as in Starlight or in an Account Transfer),
+creating the account results in an unpredictable sequence number which prevents the presigning of transactions for it. Furthermore,
+accounts can not be created and modified at the same time preventing initializing an account on behalf of another party.
+
+
+## Specification
+
+We assume a function `PublicKey derive(PublicKey pk, const *char buf, size_t
+buflen)` as specified by [ChainKD Non Hardened Extended Public
+Key](https://chain.com/docs/1.2/protocol/specifications/chainkd#derive-non-hardened-extended-public-key) which returns
```

Comment:
There's a [paper](https://cardanolaunch.com/assets/Ed25519_BIP.pdf) describing some vulnerabilities in ChainKD. Could you check if these affect this CAP in any way? Also, I couldn't find any implementations of ChainKD online.

I'm wondering if instead of deriving the key we could just make the signer field required and create an account with the `master_key = 0`, `signer[0].key = op.signer` and `signer[0].weight = 1`. This will be much easier to use as public key of a new account can be just SHA256 of tx hash and op id so no special derivation libraries are needed in stellar-core and SDKs to create smart contracts.

---

```xdr
struct CreateDeterministicAccountOp
{
    int64 startingBalance;
    Signer signer;
};
```

Reply:
> So we discussed that before and I think it was highly desired that account id continues to be a public key.

It's not the case currently as stellar-core does not check if a new account's public key (created by `create_account`) is a valid ed25519 public key.

I'm fine with ChainKD (if it's confirmed to be safe) but if there's no real advantage of using it, I'm voting for a simple hash as it's much easier option.

Reply:
Strong, strong vote in favor of using a hash rather than a public key.